Activate directories for pre-commit hook

wwmayer opened the pull request.

3x380V commented:
Hmm... A ~100k lines diff. Unreadable. Also most likely breaks queued patches. That's probably unavoidable, people will resolve conflicts manually, but can we:
- at least split this to smaller actually readable commits
- explicitly state no functional changes intended
- make each PR triggering [pre-commit-ci] unmergeable until author fixes it, so this hook does not trigger

Above would make life of those not following development closely easier as git history becomes cleaner. As people are more reading code that writing it, this could save great amount of time in total. Thank you for consideration.

wwmayer commented:
All what I have changed is: https://github.com/FreeCAD/FreeCAD/pull/16077/commits/89a6ffacf7f91d51523a5ae94189454318a9199c and the rest is automatically done by the the pre-commit hook. The PR can always be rebased easily if needed.

With several preceding PRs I have already changed the source code that needs some special treatment but sooner or later this PR must be merged because the whole point of the pre-commit hook is to have a consistent formatting of the entire code base.

3x380V commented:
Yes, it is obvious what is a change on your own and what was generated. Those several preceding PRs are doing basically the same: setting rules, then hook kicks in and does the formatting. In the ideal world, no automagically generated commits should exists. Now, if you force the reformat, all conficts in pending PRs needs to be resolved manually. The alternative would be do not let any PR with commit generated by a hook enter codebase and leave that to PR author.
The workflow is then:
- create PR as usual
- check if commit hook complaints
- fix that by squashing relevant changes to PR author changes.
- the rest of formatting changes goes to separate 'clang-format stuff' type commit
- force push those changes
- no commit hook triggers, ready to merge

That way you gradually change your code base. Not sure how much consistent formatting of the entire code base adds to the ability to easily search for changes when hunting bugs introduced during this development cycle.
I can understand this is not an issue for people working on FreeCAD on daily basis, but for occasional developers, this is a nightmare.

3x380V commented:
Okay, since more and more code is refomated, could that situation be used also to unify line endings? Use in LF repository and set `core.autocrlf` to `true` at least for C++ and Python code. When causing damage, make it worth it...

maxwxyz commented:
@wwmayer what is left to do here to be able to mark it as ready for review?

adrianinsaval commented:
Thinking out loud, if we're doing a big reformat could it make sense to include it in 1.0 to make back porting easier?

wwmayer commented:
If this tool only re-formats the code changes of a PR but not the whole file then it won't be of much help because this way we will never finish the work.

In the past I did it this way:
* Add the directory to .pre-commit-config.yaml
* Open the files of the directory that should be re-formatted into an editor and add a whitespace at the end of an arbitrary line
* Try to commit the changes
* The git commit hook will jump in and re-formats all touched files and stops with an error
* Look at the changes with git diff (or look at the whole file in an editor) and fix some bad changes manually
* Repeating the commit should succeed now

With this workflow we can re-format the code base in smaller steps.

ein-shved commented:
> If this tool only re-formats the code changes of a PR but not the whole file then it won't be of much help because this way we will never finish the work.

I agree with that, but with this tool you are able to prevent things to go worse while submitting new functionality.



> With this workflow we can re-format the code base in smaller steps.

How exactly this workflow differs from `clang-format file/to/format.cpp`?

adrianinsaval commented:
>     * Open the files of the directory that should be re-formatted into an editor and add a whitespace at the end of an arbitrary line
> 
>     * Try to commit the changes
> 
>     * The git commit hook will jump in and re-formats all touched files and stops with an error

you could just run `pre-commit run --all-file` in terminal

adrianinsaval commented:
> I would propose to fix line endings as well. It is quite annoying to see CRLF line endings on Unix and those are constantly sneaking in even in files added this year. If there is a way to prevent that, I would highly appreciate it.

I believe precommit does line endings too

wwmayer commented:
> How exactly this workflow differs from clang-format file/to/format.cpp?

Probably not at all. It's just a matter how one is faster.

ein-shved commented:
Btw, found another alternative to old clang-format-diff: [git-clang-format](https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/git-clang-format)

3x380V commented:
> Given that the utility only re-formats the part of the code that has changed but leaves the rest of the file untouched then we have to touch every line of every file until it's done.

Here 'it's done' means it is reformatted. Assumption is files which are greatly affected by reformatting need some love anyway. And this is a) opportunity to make them better, b) indication we should do so. I'm aware our opinions differs and I'm already repeating myself, so only my wish for fixing line endings stands in case of reformatting.

chennes commented:
> I believe precommit does line endings too

With our current configuration, it cleans line-endings on a per-file basis. So it makes a given file use the same endings throughout, but does not actually force either of the two specific options. This could be changed.

ein-shved commented:
Sorry, but I can not understand clearly. Are you going to run CS checker on CI? For example, [nixpkgs](https://github.com/NixOS/nixpkgs) do, and it is huge legacy project (almost ~20 years) and they did not reformat all the code (I suppose because of the `git blame` problem too) but their robots performs style checks on newly added by patches files and requires them to be well-formatted before submitting.

yorikvanhavre commented:
> it is huge legacy project (almost ~20 years)

So is FreeCAD ;) (Sorry, couldn't resist)

We've been using pre-commit for some time now and it's working rather well, usually one does not even see it... IMHO we should merge this pretty soon because it gets obsoleted very quickly and will need heavy rebasing each time

3x380V commented:
That 'usually' is a bit unusual.
```
$ git log | grep "Author: pre-commit-ci" | wc -l
47
```
These are separate commits only to deal with formatting issues. Even older projects, Linux, U-Boot, never gave reformatting a try nor autofixed anything. Patches get verified using [checkpatch](https://docs.u-boot.org/en/latest/develop/checkpatch.html) and other tools and rejected if they fail to pass. It is up to submitter to fix them, no automagic involved.
Despite I'm against reformatting code for a sake of reformatting I agree sooner this gets done, the better. Both for main and 1.0, to make backporting easier.

ein-shved commented:
HI! I have an Idea - how to reformat all the code and leave git-blame correct. Of course - you have to rewrite the history, but you can do it in separate branch an then merge it to main.

The process is fully automatic and simple.

Run rebasing to the initial formatted commit with next process for each commit
1. If we have merge conflicts - take the version of cherry-picking commit
2. Apply format to whole current code
3. Amend commit with keeping author
4. Go to next commit
